src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -576,9 +576,16 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         self.build_config.requested_target.as_ref().map(|s| &s[..])
     }
 
-    pub fn lib_profile(&self, _pkg: &PackageId) -> &'a Profile {
+    pub fn lib_profile(&self, pkg: &Package) -> &'a Profile {
         if self.build_config.release {
-            &self.profiles.release
+            return &self.profiles.release
+        }
+        if pkg.is_local() {
+            return &self.profiles.dev
+        }
+
+        if let Some(ref id) = self.build_config.deps_profile {
```

Comment:
I think this check should probably come first before the check for `--release` as the release profile may want to switch profiles as well